### ISSUE TITLE
[question] is ppo clipped predict values has gradient or it is just ok that do not mind the gradient when predict values is out of clip_range_vf boundary ?

### ISSUE THREAD

**Issue by wujiren** [DLR-RM/stable-baselines3#159]
in ppo, when predict values out of the boundary, it will be clipped as the code below:
```python
                if self.clip_range_vf is None:
                    # No clipping
                    values_pred = values
                else:
                    # Clip the different between old and new value
                    # NOTE: this depends on the reward scaling
                    values_pred = rollout_data.old_values + th.clamp(
                        values - rollout_data.old_values, -clip_range_vf, clip_range_vf
                    )
```
i just want to know if a value happened to be clipped, does it still pass the gradient to optimize, so i did a test in torch, codes are like this:
```python
import torch
from torch.nn import functional as F


class SimpleModel(torch.nn.Module):
    def __init__(self):
        super(SimpleModel, self).__init__()
        self.p = torch.nn.parameter.Parameter(torch.Tensor([1.5]), requires_grad=True)

    def forward(self, input_tensor):
        input_tensor = input_tensor * self.p
        return input_tensor


def check_gradient(input_values):
    # init the test model
    simple_model = SimpleModel()
    # get a optimizer
    optimizer = torch.optim.Adam(
        simple_model.parameters(), lr=0.1, weight_decay=0
    )
    
    # create something like ppo 
    values = simple_model(torch.Tensor([input_values]))
    old_value = torch.Tensor([1])
    returns = torch.Tensor([1])
    
    # clip the value and check the gradient
    clip_range_vf = 1
    values_pred = old_value + torch.clamp(
        values - old_value, -clip_range_vf, clip_range_vf
    )
    print(
        " values {}\n old_value {}\n returns {}\n clip_range_vf {}\n values_pred {}\n".format(
            values, old_value, returns, clip_range_vf, values_pred
        )
    )
    value_loss = F.mse_loss(returns, values_pred)
    optimizer.zero_grad()
    value_loss.backward()
    print(simple_model.p)
    optimizer.step()
    for name, parms in simple_model.named_parameters():
        print(
            "-->name:",
            name,
            "-->grad_requirs:",
            parms.requires_grad,
            " -->grad_value:",
            parms.grad,
        )
    print(simple_model.p)

``` 
when i check a input of 1 like this, the value is no clipped, and it has a gradient
```python
check_gradient(1)
```
the output is:
```
 values tensor([1.5000], grad_fn=<MulBackward0>)
 old_value tensor([1.])
 returns tensor([1.])
 clip_range_vf 1
 values_pred tensor([1.5000], grad_fn=<AddBackward0>)

Parameter containing:
tensor([1.5000], requires_grad=True)
-->name: p -->grad_requirs: True  -->grad_value: tensor([1.])
Parameter containing:
tensor([1.4000], requires_grad=True)
```
it has a gradient and the parameter changes.(to be fair to compare,i set the decay weight of adam to 0)

but when i input 2
```python
check_gradient(2)
```
the output becomes:
```
 values tensor([3.], grad_fn=<MulBackward0>)
 old_value tensor([1.])
 returns tensor([1.])
 clip_range_vf 1
 values_pred tensor([2.], grad_fn=<AddBackward0>)

Parameter containing:
tensor([1.5000], requires_grad=True)
-->name: p -->grad_requirs: True  -->grad_value: tensor([0.])
Parameter containing:
tensor([1.5000], requires_grad=True)
```
it has no gradient and the parameter remains the same.

if in a huge sample data set and with a large batch size, maybe it is ok if just some gradients is missed. but i think it could be better if we do not clip the predict value, but change the sampled data's returns to just at the boundary of old predict values, so that all  gradient could be correctly passed to the optimizer. and i think same thing happened to policy gradient.
did i make any mistake in the codes above? hope to get an answer from you.

**Comment by Miffyli** [comment]
This is done following the OpenAI Baseline's implementation where vf/pi were clipped in this fashion, and I believe it is by design (after all, the goal is to disregard updates from "too different" pi/vf functions), as demonstrated in the [PPO paper](https://arxiv.org/abs/1707.06347). Replacing the zero-gradients with something else falls into "new research" category and is thus outside stable-baselines' territory.

**Comment by araffin** [comment]
I would also add that this is disabled by default and not recommended, see [Implementation Matters in Deep RL: A Case Study on PPO and TRPO ](https://openreview.net/forum?id=r1etN1rtPB).

**Comment by wujiren** [comment]
thank you for your replies, now i think you are right.
